### ISSUE TITLE
3D sounds no longer work in GG

### ISSUE THREAD

**Issue by AmenMoses** [TheGameCreators/GameGuruRepo#1042]
Not sure when they stopped working but I think it was back in 2019 sometime.

Haven't checked MAX for the same issue but I was under the impression the sound code was shared with Classic so I assume it's broken in both.

I'll go back through the builds and see if I can pinpoint the last time it was working properly.

**Comment by AmenMoses** [comment]
Hmm, so I've gone back to Nov 2018 and my memory must be failing me because although the 3D positioning was slightly better in that build compared to the most recent, it isn't anywhere near as good as I  remember it being.
What seems to have happened between then and now is that the overall volume seems higher and there is less difference between left and right channels so the 3D effect is lessened overall.
The audio now also seems noisier somehow, same sound samples same fpm etc but a very noticeable difference in the sound.

**Comment by plemsoft** [comment]
@AmenMoses Thanks for reporting this issue , just tried a moving barrel circling around you and you can clearly hear it going from one speaker to the other while it circle , and getting longer away reduce the volume , so looks like it works fine here ?

You must call "LoopSound" continually for it to work.
Also it must be a mono wav or 3D will be disabled.

Try this little script and fpm and let me know if it dont work for you, just go into the center of  the circling barrel :)

[movewithsound.zip](https://github.com/TheGameCreators/GameGuruRepo/files/6500260/movewithsound.zip)

Thanks for your help :)

**Comment by AmenMoses** [comment]
The whole point of LoopSound is that you only call it once!

Move the LoopSound call into the _init function and it no longer works.

Lee made a change some time ago to move the sound positioning of looped sounds, has that been reverted?

Replace the LoopSound in _main with PlaySoundIfSilent( e, 0 ) and it will work but the 3D positioning is awful, stand facing the circling barrel so that it passes in front of you and then circles away, When it is on the far side of the circle it is way too loud.

It's as if the volume fall off with distance and the stereo positioning are set wrong.

Originally the sound volume for entity attached sounds was such that a value of 50 or so was virtually silent and increments after that seemed to be logarithmic, I did wonder if the value represented decibels in fact.  Then all of a sudden someone 'fixed' the sound and then we couldn't get it to be silent even at a value of 0 so it got 'fixed' again, I'm pretty sure somewhere in those 'fixes' the 3D part got screwed up.

Problem is I can't find a version of GG that works properly with my audio tutorial on the forums anymore.

**Comment by plemsoft** [comment]
@AmenMoses Yes Lee did fix this in 2008 https://github.com/TheGameCreators/GameGuruRepo/issues/35#issuecomment-371887361 , but again you need to call LoopSound constantly to update the position, actually the function start playing in the first call, all calls after that "only" set the position the sound.

LoopSound is the only way to update the sound position currently and has always been like that.

Only characters , weapons , bullets have automated sound positions.

Yes someone asked to expand the volume range :) (as it was integers) and 50 was silence, so the range was expanded, that lead to 0 was not totally silence , we then shorten the range again to be closer to the original (50 silence) and its now at around (30) silence. This seams to be a good compromise. Also this volume has nothing to do with the 3D sound, this is only a per sound max volume. The 3D sound has always been controlled 100% by the direct x sound api , we only set the position of the sound everything else is controlled by the api so there have been no changes to that ?

To get the same results as prior 2018 you could just set the volume of the entity to around "75" this way it will work just like before.

Any other commands PlaySoundIfSilent ... only set the sound position when the sound starts.

The volume control is in DB , it is set on the range of -10000 to 0 where 0 is normal and -10000 is -100db.

Im not sure what to do with this, i can change this into a feature request for a system to automate the position of the sounds if that is what your looking for ? , or perhaps just a command to set the sound position ?

**Comment by AmenMoses** [comment]
Yep, a feature request to update the sound to AAA, at least in MAX would probably be the way to go.

Because currently it is one area letting the product down.  

Way back when I did that tutorial there were work-arounds to get the sound sort of decent but recently a lot of my clients have reported that those work around are no longer working, so what used to be subtle atmospheric sounds that filled the 3D sound scape for an immersive experience are now harsh intrusive and the exact opposite of immersive.

Not sure what exactly went wrong but not having visibility of the current code base doesn't help, it would be nice to simply pile into the code myself and find the underlying cause.

btw, calling LoopSound repeatedly just to get 3D sound to sort of work is a really bad solution to what should have been a simple fix, i.e. if the object making the sound moves, move the flipping sound position!

**Comment by plemsoft** [comment]
@AmenMoses There is no changes from what you have csoundc.cpp and whats in the new repo, the only difference is the max volume, to make the old "work-arounds" sound the same as before is just to lower the volume so instead of using volume 100 just set it to 70 and it will sound just the same as before.

If you look at csoundc.cpp you see why 3D sound is not automated and not used so often, in the current setup (DirectSound) reposition a sound is really really slow, and with just a small amount of 3D real-time updated sounds - it will drop your fps.

@LeeBamberTGC 

As this is now an enhancement request i like to suggest:

Convert csoundc.cpp into using DirectXTK. https://github.com/Microsoft/DirectXTK/wiki/Audio

1. It is way faster then DirectSound.
2. It is able to use old directx driver or will if available use the builtin XAudio 2.9 on Windows 10.
3. DirectXTK is already part of GG code. 32bit and 64bit.
4. XAudio is what is used in modern xbox and windows directx games.

With the improved speed we would be able to use many more real-time 3D sounds and automate the position of sounds.

**Comment by LeeBamberTGC** [comment]
@plemsoft Sounds like a plan, and would enable us to have DX keep moving the flipping 3D sound positions with no performance cost :)  Hopefully, it opens up some extra features too which could tap into down the road.  I think good 3D sound will be even more important as some game authors switch to making VR games too!

**Comment by rvanner** [comment]
@plemsoft Is this now resolved? Can this issue be closed down?
